linalg/tensor.hpp

```diff
@@ -0,0 +1,2051 @@
+ * @note For N-by-N matrices with N > 3, requires Gaussian elimination
+ * with partial pivoting
+ */
+template <typename T, int n> MFEM_HOST_DEVICE
+tensor<T, n, n> inv(const tensor<T, n, n>& A)
+{
+   auto abs  = [](double x) { return (x < 0) ? -x : x; };
+   auto swap = [](tensor<T, n>& x, tensor<T, n>& y)
+   {
+      auto tmp = x;
+      x        = y;
+      y        = tmp;
+   };
+
+   tensor<double, n, n> B = Identity<n>();
+
+   for (int i = 0; i < n; i++)
+   {
+      // Search for maximum in this column
+      double max_val = abs(A[i][i]);
+
+      int max_row = i;
+      for (int j = i + 1; j < n; j++)
+      {
+         if (abs(A[j][i]) > max_val)
+         {
+            max_val = abs(A[j][i]);
+            max_row = j;
+         }
+      }
+
+      swap(B[max_row], B[i]);
+      swap(A[max_row], A[i]);
+
+      // zero entries below in this column
+      for (int j = i + 1; j < n; j++)
+      {
+         if (A[j][i] != 0.0)
+         {
+            double c = -A[j][i] / A[i][i];
+            A[j] += c * A[i];
+            B[j] += c * B[i];
+            A[j][i] = 0;
+         }
+      }
+   }
+
+   // upper triangular solve
+   for (int i = n - 1; i >= 0; i--)
+   {
+      B[i] = B[i] / A[i][i];
+      for (int j = i - 1; j >= 0; j--)
+      {
+         if (A[j][i] != 0.0)
+         {
+            B[j] -= A[j][i] * B[i];
+         }
+      }
+   }
+
+   return B;
+}
+
+/**
+ * @overload
+ * @note when inverting a tensor of dual numbers,
+ * hardcode the analytic derivative of the
+ * inverse of a square matrix, rather than
+ * apply gauss elimination directly on the dual number types
+ *
+ * TODO: compare performance of this hardcoded implementation to just using inv() directly
+ */
+template <typename value_type, typename gradient_type, int n> MFEM_HOST_DEVICE
+dual<value_type, gradient_type> inv(
+   tensor<dual<value_type, gradient_type>, n, n> A)
+{
+   auto invA = inv(get_value(A));
+   return make_tensor<n, n>([&](int i, int j)
+   {
+      auto          value = invA[i][j];
+      gradient_type gradient{};
+      for (int k = 0; k < n; k++)
+      {
+         for (int l = 0; l < n; l++)
+         {
+            gradient -= invA[i][k] * A[k][l].gradient * invA[l][j];
+         }
+      }
+      return dual<value_type, gradient_type> {value, gradient};
+   });
+}
+
+/**
+ * @brief recursively serialize the entries in a tensor to an output stream.
+ * Output format uses braces and comma separators to mimic C syntax for multidimensional array
+ * initialization.
+ *
+ * @param[in] out The stream to work with standard output streams
+ * @param[in] A The tensor to write out
+ */
+template <typename T, int... n>
+std::ostream& operator<<(std::ostream& out, const tensor<T, n...>& A)
+{
+   out << '{' << A[0];
+   for (int i = 1; i < tensor<T, n...>::first_dim; i++)
+   {
+      out << ", " << A[i];
+   }
+   out << '}';
+   return out;
+}
+
+/**
+ * @brief replace all entries in a tensor satisfying |x| < 1.0e-10 by literal zero
```

Comment:
Could the tolerance be provided as an optional argument rather than hardcoded constant?

Reply:
Yes, already changed in serac.

Reply:
The question is about this PR, not serac.

Actually, I don't see a good reason for this method to be included. It is better to provide a generic function that applies any given lambda function to every tensor entry.